<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta name="description" content="About Haskell humanly. A book about amazing programming language. For real programmers.">
        <meta name="author" content="Denis Shevchenko">
        <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1.0" />
        
        <link rel="icon" href="../../static/images/favicon.ico">

        <title>Handling exception</title>

        <link rel="stylesheet" href="../../static/fonts/font-awesome/css/font-awesome.min.css">

        <script src="../../static/js/jquery-2.1.4.min.js"></script>

        <!-- Compiled and minified CSS -->
        <link rel="stylesheet" href="../../static/css/materialize.min.css">

        <!-- Compiled and minified JavaScript -->
        <script src="../../static/js/materialize.min.js"></script>
        
        <!-- Own css and js -->
        <link href="../../static/css/default.css" rel="stylesheet">
        <script src="../../static/js/default.js"></script>

        <script>
          window.___gcfg = {
            lang: 'ru'
          };
        </script>

        <script src="https://apis.google.com/js/platform.js" async defer>
        </script>
    </head>

    <body>
        <div class="navbar-fixed">
            <nav>
                <div class="nav-wrapper">
                    <a href="../../en/index.html" class="brand-logo center">
                        #ohaskell
                    </a>

                    <a href="#" data-activates="mobile-demo" class="button-collapse show-on-large" style="padding-left: 15px;">
                        <!-- <i class="material-icons">toc</i> -->
                        <i class="fa fa-list-ul" style="font-size: 30px;"></i>
                    </a>

                    <ul class="side-nav" id="mobile-demo">
                        <li><a href="../../en/intro/index.html">Лирическое вступление</a></li>
                          <li><a href="../../en/intro/who.html">Кто</a></li>
                          <li><a href="../../en/intro/why.html">Почему</a></li>
                          <li><a href="../../en/intro/what-for.html">Зачем</a></li>
                          <li><a href="../../en/intro/for-whom.html">Для кого</a></li>
                        
                        <li><a href="../../en/prepare/index.html">Готовимся к работе</a></li>
                          <li><a href="../../en/prepare/create-project.html">Создаём проект</a></li>
                          <li><a href="../../en/prepare/about-modules-minimum.html">О модулях, минимум</a></li>
                          <li><a href="../../en/prepare/about-hackage.html">О Hackage</a></li>
                          <li><a href="../../en/prepare/about-sandbox.html">О песочнице</a></li>

                        <li><a href="../../en/about-haskell/index.html">Несколько слов о Haskell</a></li>
                          <li><a href="../../en/about-haskell/pure-functionality.html">Чистая функциональность</a></li>
                          <li><a href="../../en/about-haskell/types-three-foundations.html">Три кита типизации</a></li>
                          <li><a href="../../en/about-haskell/immutability.html">Неизменность данных</a></li>
                          <li><a href="../../en/about-haskell/laziness.html">Лень</a></li>
                          <li><a href="../../en/about-haskell/where.html">Где его использовать</a></li>

                        <li><a href="../../en/about-functions/index.html">О функциях</a></li>
                          <li><a href="../../en/about-functions/pure-functions.html">Чистые функции</a></li>
                          <li><a href="../../en/about-functions/lambda-functions.html">λ-функции</a></li>
                          <li><a href="../../en/about-functions/higher-order-functions.html">Функции высшего порядка</a></li>
                          <li><a href="../../en/about-functions/functional-chains.html">Функциональные цепочки</a></li>
                          <li><a href="../../en/about-functions/functions-and-operators.html">Функции и операторы</a></li>

                        <li><a href="../../en/about-lists/index.html">О списках</a></li>
                          <li><a href="../../en/about-lists/lists-at-a-glance.html">Списки &mdash; одним взглядом</a></li>
                          <li><a href="../../en/about-lists/ranges.html">Диапазоны</a></li>
                          <li><a href="../../en/about-lists/tuples.html">Кортежи</a></li>
                          <li><a href="../../en/about-lists/list-comprehension.html">Генераторы списков</a></li>

                        <li><a href="../../en/about-user-types/index.html">О пользовательских типах</a></li>
                          <li><a href="../../en/about-user-types/types-at-a-glance.html">Типы &mdash; одним взглядом</a></li>
                          <li><a href="../../en/about-user-types/about-value-constructors.html">О конструкторах значений</a></li>
                          <li><a href="../../en/about-user-types/type-context.html">Контекст типа</a></li>
                          <li><a href="../../en/about-user-types/composite-types.html">Составные типы</a></li>
                          <li><a href="../../en/about-user-types/deriving.html">Наследуемые типы</a></li>
                          <li><a href="../../en/about-user-types/own-type-classes.html">Собственные классы типов</a></li>
                          <li><a href="../../en/about-user-types/newtype.html">Новый тип</a></li>

                        <li><a href="../../en/io/index.html">Ввод и вывод</a></li>
                          <li><a href="../../en/io/functions-with-side-effects.html">Побочные эффекты</a></li>
                          <li><a href="../../en/io/IO-a.html">IO a</a></li>
                          <li><a href="../../en/io/do-imperative-world.html">do: императивный мир</a></li>
                          <li><a href="../../en/io/exceptions-handling.html">Обработка исключений</a></li>
                          <li><a href="../../en/io/own-exceptions.html">Собственные исключения</a></li>

                        <li><a href="../../en/delicious/index.html">Деликатесы</a></li>
                          <li><a href="../../en/delicious/monads-essence.html">Монады: суть</a></li>
                          <li><a href="../../en/delicious/monads-IO.html">Монады: на примере IO</a></li>
                          <li><a href="../../en/delicious/monads-practice.html">Монады: практика</a></li>
                          <li><a href="../../en/delicious/maybe.html">Может быть</a></li>
                          <li><a href="../../en/delicious/functors.html">Функторы</a></li>
                          <li><a href="../../en/delicious/applicative-functors.html">Аппликативные функторы</a></li>
                          <li><a href="../../en/delicious/transformers-first-meeting.html">Трансформеры: начало</a></li>
                          <li><a href="../../en/delicious/transformers-reader.html">Трансформеры: читатель</a></li>
                          <li><a href="../../en/delicious/transformers-writer.html">Трансформеры: писатель</a></li>
                          <li><a href="../../en/delicious/transformers-state.html">Трансформеры: состояние</a></li>
                          <li><a href="../../en/delicious/transformers-error.html">Трансформеры: ошибка</a></li>

                        <li><a href="../../en/miscellaneous/index.html">Разное</a></li>
                          <li><a href="../../en/miscellaneous/about-modules.html">О модулях</a></li>
                          <li><a href="../../en/miscellaneous/pattern-matching.html">Сравнение с образцом</a></li>
                          <li><a href="../../en/miscellaneous/map.html">Контейнеры: словарь</a></li>
                          <li><a href="../../en/miscellaneous/set.html">Контейнеры: множество</a></li>
                          <li><a href="../../en/miscellaneous/recursive-functions.html">Рекурсивные функции</a></li>
                          <li><a href="../../en/miscellaneous/about-apostrophe.html">Про апостроф</a></li>
                          <li><a href="../../en/miscellaneous/about-formatting.html">О форматировании</a></li>
                          <li><a href="../../en/miscellaneous/about-hlint.html">Про hlint</a></li>

                        <li><a href="../../en/integration/index.html">Интеграция</a></li>
                          <li><a href="../../en/integration/c-from-haskell.html">C из Haskell</a></li>

                        <li><a href="../../en/end/index.html">Заключение</a></li>
                          <li><a href="../../en/end/is-this-end.html">И что, это всё??</a></li>
                          <li><a href="../../en/end/gratitudes.html">Благодарности</a></li>
                    </ul>

                    <ul id="nav-mobile" class="right">
                        <li><a href="https://github.com/denisshevchenko/ohaskell" target="_blank"><span class="fa fa-github"></span></a></li>
                    </ul>
                </div>
            </nav>
        </div>

        <div class="container">
            <h1>Handling exception</h1>

            <p>We all aim at creating error-free software. Nevertheless, errors sometimes emerge, and therefore, we have to deal with them. Let’s discuss exceptions, that topic is especially important as many packages from Hackage contain code which throws exceptions.</p>
<p>First of all, we need module <code>Control.Exception</code>:</p>
<div class="sourceCode"><pre class="sourceCode haskell"><code class="sourceCode haskell"><span class="kw">import </span><span class="dt">Control.Exception</span></code></pre></div>
<p>That standard module is made for throwing and catching exceptions, either standard ones or our own. It’s significant that we can throw exception from either a pure function or from a function with side effect, but we can’t catch it in pure function.</p>
<h2 id="problem-with-file">Problem with file</h2>
<p>Usually we’ll be dealing with exceptions which are thrown from functions which interact with the outside world. Canonical example: we want to read the content of the non-existent file:</p>
<div class="sourceCode"><pre class="sourceCode haskell"><code class="sourceCode haskell"><span class="ot">main ::</span> <span class="dt">IO</span> ()
main <span class="fu">=</span> <span class="kw">do</span>
    fileContent <span class="ot">&lt;-</span> readFile <span class="st">&quot;Users/shevchenko/test.c&quot;</span>  <span class="co">-- Incorrect path...</span>
    putStrLn fileContent</code></pre></div>
<p>That gives us output:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">Real</span>: Users/shevchenko/test.c: openFile: does not exist (No such file or directory)</code></pre></div>
<p>Function <code>readFile</code> had thrown an exception, as it’s the only way it could report about a problem with file. Exception haven’t found any barriers on its way and was caught on application’s top level, after that a message about error was printed out and the application died on its feet.</p>
<h2 id="catching">Catching</h2>
<p>So:</p>
<div class="sourceCode"><pre class="sourceCode haskell"><code class="sourceCode haskell"><span class="kw">import </span><span class="dt">Control.Exception</span> 

<span class="ot">tryToOpenFile ::</span> FilePath <span class="ot">-&gt;</span> <span class="dt">IO</span> <span class="dt">String</span>
tryToOpenFile path <span class="fu">=</span>
    readFile path <span class="ot">`catch`</span> possibleErrors
    where
        <span class="ot">possibleErrors ::</span> <span class="dt">IOException</span> <span class="ot">-&gt;</span> <span class="dt">IO</span> <span class="dt">String</span>
        possibleErrors error <span class="fu">=</span> return <span class="fu">$</span> show error

<span class="ot">main ::</span> <span class="dt">IO</span> ()
main <span class="fu">=</span> <span class="kw">do</span>
    fileContent <span class="ot">&lt;-</span> tryToOpenFile <span class="st">&quot;Users/shevchenko/test.c&quot;</span>
    putStrLn fileContent</code></pre></div>
<p>Now we have a function <code>tryToOpenFile</code>, which opens file at given path, but does it carefully, utilizing function <code>catch</code>. As you already understood, function <code>catch</code> (defined in module <code>Control.Exception</code>) knows how to catch an exception. Consider its type:</p>
<div class="sourceCode"><pre class="sourceCode haskell"><code class="sourceCode haskell">catch<span class="ot"> ::</span> <span class="dt">Exception</span> e <span class="ot">=&gt;</span> <span class="dt">IO</span> a <span class="ot">-&gt;</span> (e <span class="ot">-&gt;</span> <span class="dt">IO</span> a) <span class="ot">-&gt;</span> <span class="dt">IO</span> a</code></pre></div>
<p>Function takes two arguments: first is an <code>IO</code>-action, second is a handling-function. If action thrown an exception (represented with polymorphic type <code>e</code>) that exception is passed to the handler.</p>
<p>Watch out: the type of exception is in the context of <code>Exception</code> class, which is defined in the same module <code>Control.Exception</code>. Any exception in our application must implement that class.</p>
<p>In order to improve readability we use function <code>catch</code> in infix notation, which keeps up with the name of our handler:</p>
<div class="sourceCode"><pre class="sourceCode haskell"><code class="sourceCode haskell">readFile path <span class="ot">`catch`</span> possibleErrors</code></pre></div>
<p>If we get an error while reading a file the corresponding exception is passed to our handler:</p>
<div class="sourceCode"><pre class="sourceCode haskell"><code class="sourceCode haskell"><span class="ot">possibleErrors ::</span> <span class="dt">IOException</span> <span class="ot">-&gt;</span> <span class="dt">IO</span> <span class="dt">String</span>
possibleErrors error <span class="fu">=</span> return <span class="fu">$</span> show error</code></pre></div>
<p>Handler takes a value of type <code>IOException</code>. In the handler’s body we stringify the resultant exception, then we take the resultant string that contains a description of an error and wrap it into an action <code>IO String</code>. In case if you are not happy with existing error message provide your own:</p>
<div class="sourceCode"><pre class="sourceCode haskell"><code class="sourceCode haskell">possibleErrors error <span class="fu">=</span> return <span class="st">&quot;Unable to open this file. Please check it.&quot;</span></code></pre></div>
<p>That message would be printed if the problem occurs.</p>
<h2 id="catch-the-other-way">Catch the other way</h2>
<p>Let’s rewrite our function:</p>
<div class="sourceCode"><pre class="sourceCode haskell"><code class="sourceCode haskell"><span class="ot">tryToOpenFile ::</span> FilePath <span class="ot">-&gt;</span> <span class="dt">IO</span> <span class="dt">String</span>
tryToOpenFile path <span class="fu">=</span>
    handle possibleErrors (readFile path)  <span class="co">-- Same thing but the other way.</span>
    where
        <span class="ot">possibleErrors ::</span> <span class="dt">IOException</span> <span class="ot">-&gt;</span> <span class="dt">IO</span> <span class="dt">String</span>
        possibleErrors error <span class="fu">=</span> return <span class="st">&quot;Aaaaa!!! Please check file.&quot;</span></code></pre></div>
<p>We replaced function <code>catch</code> with <code>handle</code>. There is absolutely no difference, except the order of arguments <code>catch</code> takes handler as a second argument, whereas <code>handle</code> takes it as first. Thus, <code>catch</code> is more readable in infix form and <code>handle</code> is more readable in simple form. So, take your choice.</p>
<h2 id="trying">Trying</h2>
<p>Standard function <code>try</code> takes different approach. For example:</p>
<div class="sourceCode"><pre class="sourceCode haskell"><code class="sourceCode haskell"><span class="kw">import </span><span class="dt">Control.Exception</span>

<span class="ot">main ::</span> <span class="dt">IO</span> ()
main <span class="fu">=</span> <span class="kw">do</span>
    result <span class="ot">&lt;-</span> try <span class="fu">$</span> readFile<span class="ot"> path ::</span> <span class="dt">IO</span> (<span class="dt">Either</span> <span class="dt">IOException</span> <span class="dt">String</span>)
    case result <span class="kw">of</span>
        <span class="dt">Left</span> exception <span class="ot">-&gt;</span> putStrLn <span class="fu">$</span> <span class="st">&quot;Fault: &quot;</span> <span class="fu">++</span> show exception
        <span class="dt">Right</span> content <span class="ot">-&gt;</span> putStrLn content
    where path <span class="fu">=</span> <span class="st">&quot;Users/dshevchenko/test.c&quot;</span></code></pre></div>
<p>Let’s line it out.</p>
<p>Here is a type of function <code>try</code>:</p>
<div class="sourceCode"><pre class="sourceCode haskell"><code class="sourceCode haskell"><span class="ot">try ::</span> <span class="dt">Exception</span> e <span class="ot">=&gt;</span> <span class="dt">IO</span> a <span class="ot">-&gt;</span> <span class="dt">IO</span> (<span class="dt">Either</span> e a)</code></pre></div>
<p>That function takes our <code>IO</code>-action and returns another <code>IO</code>-action, which returns value of standard type <code>Either e a</code>. <code>Either</code> is a type constructor designed to store either value which represents good result or value which represents bad result. Note, we explicitly specified type of value which function <code>try</code> returns.</p>
<div class="sourceCode"><pre class="sourceCode haskell"><code class="sourceCode haskell">try <span class="fu">$</span> readFile<span class="ot"> path ::</span> <span class="dt">IO</span> (<span class="dt">Either</span> <span class="dt">IOException</span> <span class="dt">String</span>)</code></pre></div>
<p>We said, “Let function <code>try</code> return us a value of type <code>Either IOException String</code>, which contains either value of type <code>IOException</code> (when something happened while reading a file) or a value of type <code>String</code> with the contents of file”.</p>
<p>Next, check if it was successful:</p>
<div class="sourceCode"><pre class="sourceCode haskell"><code class="sourceCode haskell"><span class="kw">case</span> result <span class="kw">of</span>
    <span class="dt">Left</span> exception <span class="ot">-&gt;</span> putStrLn <span class="fu">$</span> <span class="st">&quot;Fault: &quot;</span> <span class="fu">++</span> show exception
    <span class="dt">Right</span> content <span class="ot">-&gt;</span> putStrLn content</code></pre></div>
<p>Type <code>Either</code> has two constructors, <code>Left</code> and <code>Right</code>. In our case it can be displayed it this way:</p>
<pre><code>Either IOException String
       |           |
       Left        Right</code></pre>
<p>Using these constructors we can comprehend what happened. Structure <code>case-of</code> will help us. We say, “If <code>result</code> conforms to left value it’s a value of type <code>IOException</code>. Something went wrong, print an exception! And if <code>result</code> conforms to right value – we’ve got <code>String</code>. Great success, print the contents of the file.”</p>
<h2 id="in-the-pure-world">In the Pure World</h2>
<p>Sometimes we need to catch an exception which was thrown from pure function. For example:</p>
<div class="sourceCode"><pre class="sourceCode haskell"><code class="sourceCode haskell"><span class="kw">import </span><span class="dt">Control.Exception</span>

<span class="ot">main ::</span> <span class="dt">IO</span> ()
main <span class="fu">=</span> <span class="kw">do</span>
    result <span class="ot">&lt;-</span> try <span class="fu">$</span> <span class="dv">2</span> <span class="ot">`div`</span> <span class="dv">0</span><span class="ot"> ::</span> <span class="dt">IO</span> (<span class="dt">Either</span> <span class="dt">SomeException</span> <span class="dt">Integer</span>)
    case result <span class="kw">of</span>
        <span class="dt">Left</span> exception <span class="ot">-&gt;</span> putStrLn <span class="fu">$</span> <span class="st">&quot;Fault: &quot;</span> <span class="fu">++</span> show exception
        <span class="dt">Right</span> value <span class="ot">-&gt;</span> print value</code></pre></div>
<p>Here we tried to check the result of dividing 2 by zero. However, that wouldn’t pass the compilation. Function <code>try</code> expects to get <code>IO</code>-action on input, however standard function <code>div</code> is pure and returns normal number. Therefore, we need a small trick:</p>
<div class="sourceCode"><pre class="sourceCode haskell"><code class="sourceCode haskell"><span class="kw">import </span><span class="dt">Control.Exception</span>

<span class="ot">main ::</span> <span class="dt">IO</span> ()
main <span class="fu">=</span> <span class="kw">do</span>
    result <span class="ot">&lt;-</span> try <span class="fu">$</span> evaluate <span class="fu">$</span> <span class="dv">2</span> <span class="ot">`div`</span> <span class="dv">0</span>
                    <span class="ot">::</span> <span class="dt">IO</span> (<span class="dt">Either</span> <span class="dt">SomeException</span> <span class="dt">Integer</span>)
    case result <span class="kw">of</span>
        <span class="dt">Left</span> exception <span class="ot">-&gt;</span> putStrLn <span class="fu">$</span> <span class="st">&quot;Fault: &quot;</span> <span class="fu">++</span> show exception
        <span class="dt">Right</span> value <span class="ot">-&gt;</span> print value</code></pre></div>
<p>We wrapped the call to function <code>div</code> into a standard function <code>evaluate</code>. Now everything compiles, and on execution we get an expected angry message:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">Fault</span>: divide by zero</code></pre></div>
<p>The type of function <code>evaluate</code>:</p>
<div class="sourceCode"><pre class="sourceCode haskell"><code class="sourceCode haskell"><span class="ot">evaluate ::</span> a <span class="ot">-&gt;</span> <span class="dt">IO</span> a</code></pre></div>
<p>That function plays the role of adapter: it kind of turns the result into an <code>IO</code>-action which returns that result. After function <code>div</code> returned us a normal number function <code>evaluate</code> wrapped that number into an action expected by function <code>try</code>.</p>
<h2 id="so">So</h2>
<ul>
<li>Exception can be thrown from any function, but we can catch it only inside IO-function.</li>
<li>Function <code>catch</code> catches exception and passes it to the handler.</li>
<li>Function <code>handle</code> - is a mirroring of <code>catch</code>.</li>
<li>Function <code>try</code> co-operates with constructor of type <code>Either</code>. Analysing the value created by that type we can understand if there was a problem in the particular function or it went smooth.</li>
<li>In order to work with <code>try</code> we should put pure function into <code>evaluate</code>-wrapper</li>
</ul>
<p>Almost done. However, we definitely should take a look on how to create our own exceptions.</p>

<div style="padding-top: 30px;"><div>
<div class="row">
    <div class="col s6">
        <div class="left">
            <a href="../../en/io/do-imperative-world.html" class="btn waves-effect waves-light chapter-arrow">
                <span class="fa fa-angle-double-left"></span>
            </a>
        </div>
    </div>
    
    <div class="col s6">
        <div class="right">
            <a href="../../en/io/own-exceptions.html" class="btn waves-effect waves-light chapter-arrow">
                <span class="fa fa-angle-double-right"></span>
            </a>
        </div>
    </div>
</div>
<div style="padding-top: 40px;"><div>

<!--
<div>
    <script>
      (function() {
        var cx = '007697214108744450483:au850gpbupm';
        var gcse = document.createElement('script');
        gcse.type = 'text/javascript';
        gcse.async = true;
        gcse.src = (document.location.protocol == 'https:' ? 'https:' : 'http:') +
            '//cse.google.com/cse.js?cx=' + cx;
        var s = document.getElementsByTagName('script')[0];
        s.parentNode.insertBefore(gcse, s);
      })();
    </script>
    <gcse:search></gcse:search>
</div>
-->

<!--
<div>
    <div class="row">
        <form class="col s12" 
              id="searchbox_007697214108744450483:au850gpbupm" 
              action="http://www.google.com/cse">
          <div class="row">
            <div class="input-field col s6">
              <i class="material-icons prefix">search</i>

              <input value="007697214108744450483:au850gpbupm" name="cx" type="hidden"/>
              <input value="FORID:11" name="cof" type="hidden"/>

              <input id="icon_prefix" name="q" type="text"/>
              <label for="icon_prefix">Что ищем?</label>
            </div>
          </div>
        </form>
    </div>
</div>
-->

<div id="disqus_thread"></div>
<script type="text/javascript">
    /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
    var disqus_shortname = 'ohaskell'; // required: replace example with your forum shortname

    /* * * DON'T EDIT BELOW THIS LINE * * */
    (function() {
        var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
        dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
        (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    })();
</script>
<noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
 

        </div>
    </body>
</html>